wgengine/monitor/monitor_windows.go

```diff
@@ -7,273 +7,149 @@ package monitor
+func (eventMessage) ignore() bool { return false }
 
-	mu            sync.Mutex
-	lastNetChange time.Time
-	inFastPoll    bool // recent net change event made us go into fast polling mode (to detect proxy changes)
+type winMon struct {
+	logf   logger.Logf
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	mu                    sync.Mutex // guards following
+	addressChangeCallback *winipcfg.UnicastAddressChangeCallback
+	routeChangeCallback   *winipcfg.RouteChangeCallback
+	messagec              chan eventMessage
 }
 
 func newOSMon(logf logger.Logf, _ *Mon) (osMon, error) {
-	closeHandle, err := windows.CreateEvent(nil, 1 /* manual reset */, 0 /* unsignaled */, nil /* no name */)
+	m := &winMon{
+		logf: logf,
+	}
+
+	cba, err := winipcfg.RegisterUnicastAddressChangeCallback(m.unicastAddressChanged)
 	if err != nil {
-		return nil, fmt.Errorf("CreateEvent: %w", err)
+		m.logf("winipcfg.RegisterUnicastAddressChangeCallback error: %v", err)
+		return nil, err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	m := &winMon{
-		logf:            logf,
-		ctx:             ctx,
-		cancel:          cancel,
-		messagec:        make(chan messageOrError, 1),
-		pollTicker:      time.NewTicker(pollIntervalSlow),
-		closeHandle:     closeHandle,
-		goroutineDoneCh: make(chan struct{}),
+	cbr, err := winipcfg.RegisterRouteChangeCallback(m.routeChanged)
+	if err != nil {
+		cba.Unregister()
+		m.logf("winipcfg.RegisterRouteChangeCallback error: %v", err)
+		return nil, err
 	}
-	go m.awaitIPAndRouteChanges()
-	return m, nil
-}
 
-func (m *winMon) Close() error {
-	m.cancel()
-	m.pollTicker.Stop()
-	windows.SetEvent(m.closeHandle) // wakes up any reader blocked in Receive
+	m.addressChangeCallback = cba
+	m.routeChangeCallback = cbr
 
-	// Wait for awaitIPAndRouteChannges to be done, which could
-	// still be using the m.closeHandle.
-	<-m.goroutineDoneCh
+	m.ctx, m.cancel = context.WithCancel(context.Background())
 
-	windows.CloseHandle(m.closeHandle)
-	return nil
+	return m, nil
 }
 
-var errClosed = errors.New("closed")
+func (m *winMon) Close() error {
+	m.cancel()
 
-func (m *winMon) Receive() (message, error) {
-	for {
-		select {
-		case <-m.ctx.Done():
-			return nil, errClosed
-		case me := <-m.messagec:
-			return me.message, me.error
-		case <-m.pollTicker.C:
-			if m.stateChanged() {
-				m.logf("interface state changed (on poll)")
-				return pollStateChangedMessage{}, nil
-			}
-			m.mu.Lock()
-			if m.inFastPoll && time.Since(m.lastNetChange) > pollFastFor {
-				m.inFastPoll = false
-				m.pollTicker.Reset(pollIntervalSlow)
-			}
-			m.mu.Unlock()
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.addressChangeCallback != nil {
+		if err := m.addressChangeCallback.Unregister(); err != nil {
+			m.logf("addressChangeCallback.Unregister error: %v", err)
+			// TODO: Maybe we can simply log and ignore Unregister errors because
+			// the callback will be unregistered on process exit anyway?
+			return err
```

Comment:
remember this error to return, but you want to keep going and try to unregister the other one too probably.